### PR TITLE
Added friendly capturing with force modifier

### PR DESCRIPTION
**Pull request by Mailaender**
Closes https://github.com/OpenRA/OpenRA/issues/5691.

### CODE REVIEW

**Comment by GraionDilach** [comment]
Doesn't this break husks so they would require to use forcecapture?

**Comment by Micr0Bit** [comment]
ew ... i dont like this tbh ... that smells like allowing mapmodders to "sneak in" a option for trolling ... capture your allies buildings without them getting notified about this rule-change

**Comment by chrisforbes** [comment]
Either enable by default in some mod, or don't do it.

**Comment by Mailaender** [comment]
> Doesn't this break husks so they would require to use forcecapture?

No.

**Comment by Mailaender** [comment]
> Doesn't this break husks so they would require to use forcecapture?

Fixed.

**Comment by Mailaender** [comment]
As I already mentioned this is nothing new, but already supported. https://github.com/OpenRA/OpenRA/issues/5691#issuecomment-48113533

At the moment you just need

```yaml
Rules:
	^Building:
		ExternalCapturable:
			AllowAllies: true
		-EngineerRepair:
```

for it to work. This fixes it.

**Comment by Micr0Bit** [comment]
by supported i mean , "the idea of it getting obvious positive reactions" ... which isnt the case

**Comment by abcdefg30** [comment]
Wow, why so offensive against this change guys?

@Micr0Bit I think there are already enough ways you can troll with custom maps... And you already get the "This map contains custom rules. Game experience may change." warning when someone selects such a map, so if you don't want to risk getting trolled, don't play on those maps... I could easily create a map that lets rifle man explode (nuke explosion) when they stay on the same cell or add a script that gives players named `abcdefg30` infinite mammoth paradrops, just as example that trolling is already possible in many ways.

> Either enable by default in some mod, or don't do it.

@chrisforbes what do you mean by that? If I didn't missunderstood it: Since we do we require explicit usecases for fixing bugs/adding features?

---------------------------------------------------------------------------------------------------------------------------------------

So, somehow I don't get your points, could you please elaborate a bit more?

**Comment by Micr0Bit** [comment]
this notification ... "This map contains custom rules." doesnt work 100%

**Comment by abcdefg30** [comment]
Then file a bug report, please, so that this gets fixed.

**Comment by Micr0Bit** [comment]
still this doesnt look like it would close #5691 ... as xanax (and myself in the origin of the idea #5739 ) were looking for something to (atleast) turn on/off ingame 
something that extends the core-gameplay (worse enough now that im re-thinking it ) ... not adding another map-modder option

**Comment by Mailaender** [comment]
@pchote already rejected the lobby option variant. https://github.com/OpenRA/OpenRA/issues/5691#issuecomment-51006790

**Comment by Micr0Bit** [comment]
okay thats a point , then im indifferent on this

**Comment by GraionDilach** [comment]
I'd prefer if this would have a notation like what `Armament` got, by having a `TargetModifier` and a `ForceTargetModifier`. This way seems arbitrary atm.

**Comment by pchote** [comment]
> I'd prefer if this would have a notation like what Armament got, by having a TargetModifier and a ForceTargetModifier. This way seems arbitrary atm.

This sounds like the best approach to me.  We can also set up infiltration/repair/etc the same way.

**Comment by MustaphaTR** [comment]
> the notification ... "This map contains custom rules." doesnt work 100%

I think this notification should have a clickable link to resource center, so people can easily see what the map changes.

**Comment by Micr0Bit** [comment]
@MustaphaTR : people want to "play games" ... and not every single player to study the rules before they decide to actually stay and play on a server ... "contains custom rules" is really easy and good enough ...

**Comment by Mailaender** [comment]
> > I'd prefer if this would have a notation like what Armament got, by having a TargetModifier and a ForceTargetModifier. This way seems arbitrary atm.

> I agree (except for naming: `TargetStances` and `ForceTargetStances` like Armament would be better).  We can also set up infiltration/repair/etc the same way.

Updated as suggested.

File: OpenRA.Mods.Common/Traits/Capturable.cs [hunk 1]
```diff
@@ -32,16 +33,6 @@ public bool CanBeTargetedBy(Actor captor, Player owner)
 			if (c == null)
 				return false;
 
-			var playerRelationship = owner.Stances[captor.Owner];
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Capturable.cs]: You need to check the stances here too (and likewise for the other traits).  The order targeters are client side code, so the checks there could be modified or removed by somebody malicious.

File: OpenRA.Mods.Common/Traits/Capturable.cs [hunk 2]
```diff
@@ -17,9 +17,10 @@ public class CapturableInfo : ITraitInfo
 	{
 		[Desc("Type listed under Types in Captures: trait of actors that can capture this).")]
 		public readonly string Type = "building";
-		public readonly bool AllowAllies = false;
-		public readonly bool AllowNeutral = true;
-		public readonly bool AllowEnemies = true;
+
+		public readonly Stance TargetStances = Stance.Enemy | Stance.Neutral;
+		public readonly Stance ForceTargetStances = Stance.Enemy | Stance.Neutral | Stance.Ally;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Capturable.cs]: This sneaks in enabling for allies by default.

File: OpenRA.Mods.Common/Traits/ExternalCapturable.cs [hunk 3]
```diff
@@ -17,9 +17,10 @@ public class ExternalCapturableInfo : ITraitInfo
 	{
 		[Desc("Type of actor (the ExternalCaptures: trait defines what Types it can capture).")]
 		public readonly string Type = "building";
-		public readonly bool AllowAllies = false;
-		public readonly bool AllowNeutral = true;
-		public readonly bool AllowEnemies = true;
+
+		public readonly Stance TargetStances = Stance.Enemy | Stance.Neutral;
+		public readonly Stance ForceTargetStances = Stance.Enemy | Stance.Neutral | Stance.Ally;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/ExternalCapturable.cs]: Likewise here.

File: OpenRA.Mods.Common/Traits/Capturable.cs (hunk 1 above)
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Capturable.cs]: I added the checks back (below).

**Comment by Mailaender** [comment]
Updated.

File: OpenRA.Mods.Common/Traits/Capturable.cs [hunk 4]
```diff
@@ -33,17 +34,11 @@ public bool CanBeTargetedBy(Actor captor, Player owner)
...
-
-			if (playerRelationship == Stance.Enemy && !AllowEnemies)
-				return false;
-
-			if (playerRelationship == Stance.Neutral && !AllowNeutral)
+			if (!c.CaptureTypes.Contains(Type))
 				return false;
 
-			if (!c.CaptureTypes.Contains(Type))
+			var stances = ForceTargetStances | TargetStances;
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Capturable.cs]: This worries me a bit - in teamgame, this allow a player to capture a freshly-captured-by-his-ally actor just by arriving late.

File: OpenRA.Mods.Common/Traits/ExternalCapturable.cs [hunk 5]
```diff
@@ -33,17 +34,11 @@ public bool CanBeTargetedBy(Actor captor, Player owner)
...
-
-			if (playerRelationship == Stance.Enemy && !AllowEnemies)
-				return false;
-
-			if (playerRelationship == Stance.Neutral && !AllowNeutral)
+			if (!c.CaptureTypes.Contains(Type))
 				return false;
 
-			if (!c.CaptureTypes.Contains(Type))
+			var stances = ForceTargetStances | TargetStances;
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/ExternalCapturable.cs]: Same here.

**Comment by GraionDilach** [comment]
Could these be changed via including the force modifier passed an argument then checking for it explicitly?

:+1: otherwise.

**Comment by Mailaender** [comment]
I don't have access to the force modifier in those function, otherwise I would have mirrored the checks.

**Comment by pchote** [comment]
You can/should do what the attack code does.

**Comment by Mailaender** [comment]
I already checked. It is called in other contexts as well where those modifiers simply aren't available.

**Comment by pchote** [comment]
The two contexts I am aware of are the order targeting and ResolveOrder.  The modifiers are available in both.

**Comment by GraionDilach** [comment]
The `CaptureActor` activity also uses it inside the `CanReserve` call which sounds like the denying context - although I'm not certain what that supposed to check for there.

**Comment by pchote** [comment]
That activity will be called via ResolveOrder, so there should be no problem.

**Comment by pchote** [comment]
However it might pay to try that in a separate commit - if the plumbing ends up being horrible then the current solution might be acceptable if we change Attack to use similar logic.

File: OpenRA.Mods.Common/Traits/Capturable.cs (hunk 4 above)
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Capturable.cs]: The target generation increment that happens on owner change should stop that.

**Comment by Mailaender** [comment]
> That activity will be called via ResolveOrder, so there should be no problem.

Sorry, I can't find any `TargetModifier` near `ResolveOrder` in `Captures`.

**Comment by pchote** [comment]
While i'm no longer convinced that this is the right approach (see my last comment), my original comment on the topic said:

> You can/should do what the attack code does, which gives you access to the force modifier in those functions.

Please don't remove the changes requested tag until the discussion prompting it is resolved.  It was added in this case because of the inconsistency this introduces compared with AttackBase.

File: OpenRA.Mods.Common/Traits/EngineerRepair.cs [hunk 6]
```diff
@@ -98,40 +101,6 @@ public void ResolveOrder(Actor self, Order order)
 			self.SetTargetLine(target, Color.Yellow);
 			self.QueueActivity(new RepairBuilding(self, target.Actor, info.EnterBehaviour));
 		}
-
-		class EngineerRepairOrderTargeter : UnitOrderTargeter
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/EngineerRepair.cs]: Like discussed in other PRs please keep these as internal child classes.

**Comment by Mailaender** [comment]
Updated and experimented a bit. Exposing `TargetModifiers` into activities won't work and doesn't seem like something we want. I have no idea what to change in `AttackBase`.

**Comment by abcdefg30** [comment]
So, what is the status here?

**Comment by Mailaender** [comment]
I don't know how to fulfill the requested changes. :(

**Comment by reaperrr** [comment]
Closing for now due to lack of progress, to clean up the PR queue.

**Pull request closed without merging**